resources/2023-11-application-diagnostics-prototype.md

```diff
@@ -0,0 +1,215 @@
+- Radius can automatically enrich diagnostics information with meaningful per-component metadata that helps users navigate, understand, and search. We want to validate that app-graph information is useful for this purpose.
+- We also want to learn about the support/recommended configurations of OpenTelemetry so we can better understand how to design these features, and what level of extensibility is required in Radius. 
+
+### Non goals
+
+There's a number of things we can't solve inside Radius if we remain committed to our principles. For example Radius doesn't change the way developers write code. We also don't try to change the way operators do common tasks like provisioning non-developer-facing infrastructure.
+
+- Out-of-scope: We don't simplify the creation of diagnostics stores like Prometheus and Grafana. We assume operators create these out-of-band from Radius and provide Radius with the relevant configuration.
+- Out-of-scope: We don't auto-instrument the developer's code. Developers will still need to use libraries like the OpenTelemetry SDK to instrument their code for tracing and metrics. 
+- Out-of-scope: Because it's a prototype, there will be gaps in this end-to-end, probably due to things we haven't built yet like the dashboard. That's expected for now. 
+
+### User scenarios (optional)
+
+*These three scenarios will work together to form our prototype. These are more specific than usual, because we're going to build a specific demo.*
+
+#### Configuring OpenTelemetry as part of an environment
+
+As an operator, I can configure diagnostics stores for use in environments. Then I can configure a Radius environment to use those those diagnostics stores. When I do this I am accomplishing two things: 1) making the supported diagnostics stores and their configuration explicit as part of the environment, 2) making the configuration of collectors for applications automatic based on the configuration that I provided.
+
+In our demo setup we're going to use Azure Monitor + Grafana for both metrics, and logs. We're not addressing tracing in this demo. These technologies are good choices because Grafana is self-hostable and widely used, and Azure Monitor as a backend is a good subtitute for any OTEL-compatible source. It would be easy for us to move this demo to another cloud by swapping out Azure Monitor.
+
+#### Collecting and storing diagnostics data automatically for applications
+
+As a developer, I can instrument my code using the OpenTelemetry SDK to provide metrics. I can write logs to the console using the logging packages I prefer for my application. When I deploy my application, using Radius, the configuration for OpenTelemetry will be automatically injected. I don't know have to know about the configuration of, or have access to where the diagnostics data goes.
```

Comment:
Does OpenTelemetry have a standard environment variable or way to configure itself? Not familiar with how it's usually setup

Reply:
For the SDKs yes. This section refers to **both**:

- Setting the standard env-var that the SDKs understand
- Managing the standard collector (in one of the **many** ways to use the collector, with options for configuration)